TooltipDataCallback to allow modders add their own tooltips to existing items

JumperOnJava opened the pull request:
I recently was developing feature that added "sell" and "purchase" tooltips for blocks with that functionality. The problem is that these toolips were added by different mods, one for selling, one for purchasing. And to make these mods compatible i had to develop small separate library, and package it to both mods. And i haven't even tested compatibility with other mods that may change tooltips.
The only tooltip related events I found are for adding text lines to tooltip and registering tooltip component.
So I propose new event the Fabric API - TooltipDataCallback(ItemStack, List<TooltipData>).
Also i added three example mods to showcase what this can add [(Fabric discord message link)](https://discord.com/channels/507304429255393322/566276937035546624/1170469767669350450)

JumperOnJava committed:
MultiTooltipApi

JumperOnJava committed:
updated javadocs

JumperOnJava committed:
added license headers

JumperOnJava committed:
fixed code style

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/TooltipDataCallback.java:
This class needs top-level javadoc and javadoc on the `getTooltipData` method explaining what to do with the list.

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/TooltipDataCallback.java:
```diff
@@ -0,0 +1,41 @@
+			callback.getTooltipData(itemStack, tooltipDataList);
+		}
+	});
+	void getTooltipData(ItemStack itemStack, List<TooltipData> tooltipDataList);
```
A better name for this method would be `appendTooltipData` to match with `Item.appendTooltip` etc.

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/tooltip/MultiTooltipComponent.java:
```diff
@@ -0,0 +1,97 @@
+		int position = 0;
+
+		for (TooltipComponent c : components) {
+			matrix.translate(0, position, 0);
+			c.drawText(textRenderer, x, y, matrix, vertexConsumers);
+			matrix.translate(0, -position, 0);
+			position += c.getHeight();
```
It'd be cleaner to simply translate the matrix stack by `(0, c.getHeight(), 0)` each time, no need to translate back and forth.

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/tooltip/MultiTooltipComponent.java:
```diff
@@ -0,0 +1,97 @@
+	public void drawText(TextRenderer textRenderer, int x, int y, Matrix4f matrix, VertexConsumerProvider.Immediate vertexConsumers) {
+		int position = 0;
+
+		for (TooltipComponent c : components) {
```
This is missing push/pop guards around the matrix stack manipulation.

JumperOnJava replied:
There is only Matrix4f in this method, no stacks at all

JumperOnJava replied:
![image](https://github.com/FabricMC/fabric/assets/59884606/e4ca60e2-9abc-4ac3-a540-58bcb7e9c8e1)

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/tooltip/MultiTooltipComponent.java:
```diff
@@ -0,0 +1,97 @@
+		int position = 0;
+
+		for (TooltipComponent c : components) {
+			context.getMatrices().push();
+			context.getMatrices().translate(0, position, 0);
+			c.drawItems(textRenderer, x, y, context);
+			context.getMatrices().pop();
+			position += c.getHeight();
```
Same comment about the translation here.

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/HandledScreenMixin.java:
```diff
@@ -0,0 +1,47 @@
+			return Optional.empty();
+		}
+
+		return Optional.of(mutlidata);
```
This could just return the unwrapped data for a single child, currently this wraps *all tooltip data of any item with one* into the list variety.

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/HandledScreenMixin.java:
```diff
@@ -0,0 +1,47 @@
+	@Redirect(method = "drawMouseoverTooltip", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;getTooltipData()Ljava/util/Optional;"))
+	Optional<TooltipData> addMultiData(ItemStack stack) {
+		Optional<TooltipData> original = stack.getTooltipData();
+		var mutlidata = new MultiTooltipData(1);
```
Typo

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/tooltip/MultiTooltipData.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * This class stores multiple TooltipData object to their further mapping to MultiTooltipComponent.
+ */
+public class MultiTooltipData extends ArrayList<TooltipData> implements TooltipData {
```
Extending ArrayList feels like an unusual design pattern, something like this + a separate array list passed to it  would probably be more conventional:
```java
public record MultiTooltipData(List<TooltipData> children) implements TooltipData {}
```

JumperOnJava replied:
Yeah, i was concerned about this too, thanks for the solution

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/TooltipComponentCallback.java:
```diff
@@ -27,7 +27,8 @@
  * Allows registering a mapping from {@link TooltipData} to {@link TooltipComponent}.
- * This allows custom tooltips for items: first, override {@link Item#getTooltipData} and return a custom {@code TooltipData}.
+ * This allows custom tooltips for items: first, override {@link Item#getTooltipData} and return a custom {@code TooltipData}, or use
+ * {@link TooltipDataCallback} to add {@code TooltipData} to an existing item
```
```suggestion
 * {@link TooltipDataCallback} to add {@code TooltipData} to an existing item.
```

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/TooltipDataCallback.java:
```diff
@@ -0,0 +1,41 @@
+	/**
+	 * Allows registering custom {@link TooltipData} object for item.
+	 * This allows you to add your own tooltips to existing items.
+	 * Custom {@code TooltipData} should be registered using {@link TooltipComponentCallback},
+	 * otherwise game will crash when trying to map {@code TooltipData} to {@code TooltipComponent}
```
```suggestion
	 *
	 * <p>Custom {@link TooltipData} should be registered using {@link TooltipComponentCallback},
	 * otherwise game will crash when trying to map {@link TooltipData} to {@link TooltipComponent}.
```
Formatting + these should be links so they get remapped properly by Mercury.

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/HandledScreenMixin.java:
```diff
@@ -0,0 +1,47 @@
+
+@Mixin(HandledScreen.class)
+class HandledScreenMixin {
+	@Redirect(method = "drawMouseoverTooltip", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;getTooltipData()Ljava/util/Optional;"))
```
I'd prefer a `@ModifyArg` on the receiving method instead (but neither of them is all that compatible compared to MixinExtras injectors)

JumperOnJava committed:
Update fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/TooltipComponentCallback.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

JumperOnJava committed:
Update fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/TooltipDataCallback.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

JumperOnJava committed:
fixed issues in code found by @Juuxel

JumperOnJava committed:
Style fixes

JumperOnJava committed:
fixed tooltip data not adding

It happened when only one tooltip data is added by event handlers
I for some reason assumed that there is always original data in item

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/HandledScreenMixin.java:
```diff
@@ -34,14 +35,18 @@ class HandledScreenMixin {
 
-		if (mutlidata.size() == 0) {
-			return Optional.empty();
+		if (multiData.tooltipData().size() == 0) {
```
```suggestion
		if (multiData.tooltipData().isEmpty()) {
```

JumperOnJava committed:
Update fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/HandledScreenMixin.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 reviewed (approved):
Looks good, could possibly use Mixin extras in the Mixin, but its not a major issue.

LlamaLad7 commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/HandledScreenMixin.java:
```diff
@@ -0,0 +1,52 @@
+
+@Mixin(HandledScreen.class)
+class HandledScreenMixin {
+	@Redirect(method = "drawMouseoverTooltip", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;getTooltipData()Ljava/util/Optional;"))
```
A WrapOperation would be better.

JumperOnJava replied:
Mixin extras avaliable only on loader 0.15+, this version is 0.14 compatible
Changing it for one small feature doesn't make much sense (for me)
Here is ME version for 1.20.4 https://github.com/FabricMC/fabric/pull/3486

modmuss50 commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/HandledScreenMixin.java:
```diff
@@ -0,0 +1,52 @@
+		original.ifPresent(multiData.tooltipData()::add);
+		TooltipDataCallback.EVENT.invoker().appendTooltipData(stack, multiData.tooltipData());
+
+		if (multiData.tooltipData().isEmpty()) {
+			return original;
+		}
```
Do we expect people to be allowed to remove the original?

modmuss50 commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/HandledScreenMixin.java:
```diff
@@ -0,0 +1,52 @@
+	@Redirect(method = "drawMouseoverTooltip", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;getTooltipData()Ljava/util/Optional;"))
+	Optional<TooltipData> addMultiData(ItemStack stack) {
+		Optional<TooltipData> original = stack.getTooltipData();
+		var multiData = new MultiTooltipData(new ArrayList<>());
```
Nit: No need to create MultiTooltipData here, just start with the list, and then create it later if the list > 1

modmuss50 commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/tooltip/MultiTooltipComponentRegister.java:
```diff
@@ -0,0 +1,33 @@
+import net.fabricmc.api.ClientModInitializer;
+import net.fabricmc.fabric.api.client.rendering.v1.TooltipComponentCallback;
+
+public class MultiTooltipComponentRegister implements ClientModInitializer {
```
`MultiTooltipComponentInitializer` maybe, this class name is a little weird.

modmuss50 commented:
Remvoed from last call, there are breaking changes between the two PRs, it might just be best to have one and make sure that all the behaviour is tested/defined.
